### PR TITLE
CWL runner example

### PR DESCRIPTION
**Pull request by jbingham**
Added example to run cwltool on a single VM.

### CODE REVIEW

File: cwl_runner/README.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# Run a CWL workflow
+
+## This example demonstrates running a multi-stage workflow on Google Cloud Platform
+
+* The workflow is launched with a bash script that calls the [Google Cloud SDK](https://cloud.google.com/sdk)
```
Comment (mbookman) [inline cwl_runner/README.md]: "calls the gcloud command-line tool that is included in the Google Cloud SDK"

Comment (mbookman) [inline cwl_runner/README.md]: ... with a bash script, [cwl_runner.sh](cwl_runner.sh), that calls...

File: cwl_runner/README.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+* The workflow is defined using the [Common Workflow Language (CWL)](http://www.commonwl.org)
+* The workflow stages are orchestrated using [cwltool](https://github.com/common-workflow-language/cwltool/tree/master/cwltool)
+
+Here's what the bash script does to run a CWL workflow:
```
Comment (mbookman) [inline cwl_runner/README.md]: There's a "tense" disconnect between the intro here and the steps.
As is, the first bullet should be "Create*s* a disk".
To just update the intro and not the bullets, perhaps make it:

To run a CWL workflow, cwl_runner.sh will:

File: cwl_runner/README.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+1. Create a disk
+1. Create a Compute Engine VM with that disk
+1. Run a startup script on the VM
+
+The startup script running on the VM will:
```
Comment (mbookman) [inline cwl_runner/README.md]: The startup script, [cwl_startup.sh](cwl_startup.sh) on the VM will:

File: cwl_runner/README.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+1. Mount and format the disk
+1. Download input files from Google Cloud Storage
+1. Install Docker
+1. Install cwltool
+1. Run the CWL workflow and wait until completion
+1. Copy output files to Google Cloud Storage
+1. Copy stdout and stderr logs to Google Cloud Storage
+1. Shutdown and delete the VM and disk
+
+__Note that the CWL runner does not use the Pipelines API. If you don't have enough quota, the script will fail; it won't be queued to run when quota is available.__
```
Comment (mbookman) [inline cwl_runner/README.md]: __Note that the CWL runner does not use the [Pipelines API](https://cloud.google.com/genomics/reference/rest/v1alpha2/pipelines). If you don't have enough quota, the script will fail; it won't be queued to run when quota is available.__

File: cwl_runner/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+## Prerequisites
+
+1. Download the one required file, `cwl_runner.sh`, or, if you prefer, clone or fork this github repository.
+1. Enable the Genomics, Cloud Storage, and Compute Engine APIs on a new or existing Google Cloud Project using the [Cloud Console](https://console.cloud.google.com/flows/enableapi?apiid=storage_component,compute_component&redirect=https://console.cloud.google.com)
+1. Install and initialize the [Google Cloud SDK](https://cloud.google.com/sdk).
+1. Follow the Cloud Storage instructions for [Creating Storage Buckets](https://cloud.google.com/storage/docs/creating-buckets) to create a bucket to store workflow output and logging
+
+## Running a sample workflow in the cloud
+
+This script can (in theory!) run any CWL workflow.
```
Comment (mbookman) [inline cwl_runner/README.md]: How about:

This script should be able to support any workflow supported by cwltool.

File: cwl_runner/README.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+1. Run the CWL workflow and wait until completion
+1. Copy output files to Google Cloud Storage
+1. Copy stdout and stderr logs to Google Cloud Storage
+1. Shutdown and delete the VM and disk
+
+__Note that the CWL runner does not use the Pipelines API. If you don't have enough quota, the script will fail; it won't be queued to run when quota is available.__
+
+## Prerequisites
+
+1. Download the one required file, `cwl_runner.sh`, or, if you prefer, clone or fork this github repository.
```
Comment (mbookman) [inline cwl_runner/README.md]: The lines in cwl_runner.sh:

    gsutil cp "${STARTUP_SCRIPT}" "${STARTUP_SCRIPT_URL}"
    gsutil cp "${SHUTDOWN_SCRIPT}" "${SHUTDOWN_SCRIPT_URL}"

mean that the startup script and shutdown script need to be copied down as well.

File: cwl_runner/README.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+You can run the script with `--help` to see all of the command-line options.
+
+```
+./cwl_runner.sh --help
+```
+
+As an example, let's run a real workflow from the [Genome Data Commons](https://gdc.cancer.gov) stored in a [GDC github project](https://github.com/nci-gdc/gdc-dnaseq-cwl).
+
+This particular workflow requires a reference genome bundle, a DNA reads file in BAM format, and several CWL tool definitions. All of those files have already been copied into a Cloud Storage bucket, so we can just reference them when we run the CWL workflow.
```
Comment (mbookman) [inline cwl_runner/README.md]: This particular workflow requires:

* a reference genome bundle
* a DNA reads file in BAM format
* several CWL tool definitions

All of those files have already been copied into Google Cloud Storage (at gs://genomics-public-data/cwl-examples/gdc-dnaseq-cwl)...

File: cwl_runner/README.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+Once your job starts, it will have an `OPERATION-ID` assigned, which you can use to check status and find the VM and disk in your cloud project.
+
+To monitor your job, check the status to see if it's RUNNING, COMPLETED, or FAILED:
+```
+gsutil cat gs://MY-BUCKET/MY-PATH/status-OPERATION-ID.txt
+```
+
+While your job is running, you can see the VM in the cloud console and command-line.
+When the job completes, the VM will no longer be found unless --keep-alive is set.
```
Comment (mbookman) [inline cwl_runner/README.md]: `--keep-alive`

File: cwl_runner/README.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+Command-line:  
+```
+gcloud compute instances describe cwl-vm-OPERATION-ID
+```
+
+## Canceling a job
+
+To cancel a running job, you can terminate the VM from the cloud console or command-line:
```
Comment (mbookman) [inline cwl_runner/README.md]: [Cloud Console](https://console.cloud.google.com/compute/instances)

File: cwl_runner/cwl_runner.sh [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# cwl_runner.sh
+#
+# From your shell prompt, launch a Google Compute Engine VM to run
+# a Common Workflow Language (CWL) workflow with cwltool.
+
+WORKFLOW_PATH=
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: declare WORKFLOW_PATH=

<etc>

File: cwl_runner/cwl_runner.sh [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+  Run with a preemptible VM that costs less but may be terminated before finishing.
+-r --runner STRING
+  The CWL runner to use. Values can be "cwltool" or "rabix". Default: ${RUNNER}.
+-z --zone STRING
+  The zone to launch the VM and disk in. If omitted, your default project zone will be used.
+
+EOM
+
+# Parse command-line
+while [[ $# -gt 0 ]]
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: while [[ $# -gt 0 ]]; do

File: cwl_runner/cwl_runner.sh [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+  The CWL runner to use. Values can be "cwltool" or "rabix". Default: ${RUNNER}.
+-z --zone STRING
+  The zone to launch the VM and disk in. If omitted, your default project zone will be used.
+
+EOM
+
+# Parse command-line
+while [[ $# -gt 0 ]] 
+do
+key="$1"
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: Indent the contents of the `while` loop.

File: cwl_runner/cwl_runner.sh [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+  shift
+  ;;
+  *)
+  # unknown option
+  ;;
+esac
+shift
+done
+
+set -e errexit
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: Is there a reason not to do this earlier?
This should be "set -o errexit"
Should also include "set -o nounset"

File: cwl_runner/cwl_runner.sh [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+if [[ -z "${WORKFLOW_FILE}" || -z "${SETTINGS_FILE}" || -z "${OUTPUT}" ]]; then
+  echo "Error: Missing required argument(s)."
+  echo "${HELP_MESSAGE}"
+  exit 1
+fi
+
+DISK_NAME="cwl-disk-${OPERATION_ID}"
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: readonly DISK_NAME="cwl-disk-${OPERATION_ID}"
    readonly DISK_CMD="gcloud compute disks create ${DISK_NAME} ${ZONE} --size ${DISK_SIZE}"

File: cwl_runner/cwl_runner.sh [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+DISK_CMD="gcloud compute disks create ${DISK_NAME} ${ZONE} --size ${DISK_SIZE}"
+
+PWD="$( cd $( dirname ${BASH_SOURCE[0]} ) && pwd )"
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: "PWD" is an environment variable that should already be set. I think you want to rename this `SCRIPT_DIR` or some such.

File: cwl_runner/cwl_runner.sh [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+
+STARTUP_SCRIPT_NAME="cwl_startup.sh"
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: Mark these all "readonly".

File: cwl_runner/cwl_runner.sh [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+SHUTDOWN_SCRIPT_NAME="cwl_shutdown.sh"
+SHUTDOWN_SCRIPT="${PWD}/${SHUTDOWN_SCRIPT_NAME}"
+SHUTDOWN_SCRIPT_URL="${OUTPUT}/${SHUTDOWN_SCRIPT_NAME%.*}-${OPERATION_ID}.sh"
+
+STATUS_FILE="${OUTPUT}/status-${OPERATION_ID}.txt"
+STATUS="STARTING"
+
+VM_NAME="cwl-vm-${OPERATION_ID}"
+VM_CMD="gcloud compute instances create ${VM_NAME} \
+--disk name=${DISK_NAME},device-name=${DISK_NAME},auto-delete=yes \
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: Indent all of the command flags.
It inserts spaces into the string, but it does not matter and reads better.

File: cwl_runner/cwl_runner.sh [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+input=\"${INPUT}\",\
+input-recursive=\"${INPUT_RECURSIVE}\",\
+output=${OUTPUT},\
+runner=${RUNNER},\
+status-file=${STATUS_FILE},\
+keep-alive=${KEEP_ALIVE}"
+
+echo $(date)
+echo "Generating script commands and writing to file"
+TMP_SCRIPT=".$(basename ${0%.*} )-${OPERATION_ID}.sh"
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: readonly

File: cwl_runner/cwl_runner.sh [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+echo "#!/bin/bash" > "${TMP_SCRIPT}"
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]:
```
cat > "${TMP_SCRIPT}" << EOF
#!/bin/bash
echo "${DISK_CMD}"
echo "${VM_CMD}"
EOF
```

File: cwl_runner/cwl_runner.sh [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+echo "$VM_CMD" >> "${TMP_SCRIPT}"
+
+echo "Copying scripts to the output path in Cloud Storage"
+gsutil cp "${STARTUP_SCRIPT}" "${STARTUP_SCRIPT_URL}"
+gsutil cp "${SHUTDOWN_SCRIPT}" "${SHUTDOWN_SCRIPT_URL}"
+gsutil cp "${TMP_SCRIPT}" "${OUTPUT}/${TMP_SCRIPT/./}"
+rm "${TMP_SCRIPT}"
+
+echo "Creating Google Compute Engine VM and disk"
+echo $DISK_CMD
```
Comment (mbookman) [inline cwl_runner/cwl_runner.sh]: Wrap DISK_CMD and VM_CMD in curly braces.

File: cwl_runner/cwl_shutdown.sh [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+# limitations under the License.
+
+# cwl_shutdown.sh
+#
+# This is the shutdown script that runs on Compute Engine before the VM shuts down.
+
+echo "$(date)"
+echo "Running shutdown script"
+
+METADATA_URL="http://metadata.google.internal/computeMetadata/v1/instance"
```
Comment (mbookman) [inline cwl_runner/cwl_shutdown.sh]: Mark these variables `readonly`

File: cwl_runner/cwl_shutdown.sh [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+OPERATION_ID=$(curl "${METADATA_URL}/attributes/operation-id" -H "${METADATA_HEADERS}")
+STATUS_LOCAL="/tmp/status-${OPERATION_ID}.txt"
+
+STDOUT=/tmp/stdout-${OPERATION_ID}.txt
+STDERR=/tmp/stderr-${OPERATION_ID}.txt
+
+echo "$(date)"
+echo "Copying stdout and stderr to Cloud Storage"
+CMD="gsutil -m cp ${STDOUT} ${STDERR} ${OUTPUT}/"
+echo $CMD
```
Comment (mbookman) [inline cwl_runner/cwl_shutdown.sh]: Wrap the references to CMD in curly braces.

File: cwl_runner/cwl_shutdown.sh [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+$CMD
+
+# Typically shutdown will cause a running job to fail and status will be set to FAILED
+# In case the status is left as RUNNING, set it to FAILED
+STATUS=$(cat ${STATUS_LOCAL})
+echo "Status ${STATUS}"
+
+if [[ ${STATUS} == "RUNNING" ]]; then
```
Comment (mbookman) [inline cwl_runner/cwl_shutdown.sh]: Double-quotes around "${STATUS}"

File: cwl_runner/cwl_startup.sh [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# cwl_startup_script.sh
+#
+# This is the startup script that runs on Compute Engine to run
+# a Common Workflow Language (CWL) workflow with cwltool.
+
+METADATA_URL="http://metadata.google.internal/computeMetadata/v1/instance"
```
Comment (mbookman) [inline cwl_runner/cwl_startup.sh]: Mark these variables `readonly`

File: cwl_runner/cwl_startup.sh [hunk 25]
```diff
@@ -0,0 +1,197 @@
...
+INPUT_FOLDER="${MOUNT_POINT}/input"
+OUTPUT_FOLDER="${MOUNT_POINT}/output"
+TMP_FOLDER="${MOUNT_POINT}/tmp"
+sudo mkdir -m 777 -p "${INPUT_FOLDER}"
+sudo mkdir -m 777 -p "${OUTPUT_FOLDER}"
+sudo mkdir -m 777 -p "${TMP_FOLDER}"
+
+echo "$(date)"
+echo "Copying input files to local disk"
+while IFS=';' read -ra URL; do
```
Comment (mbookman) [inline cwl_runner/cwl_startup.sh]: Change `URL` to `URL_LIST`.

File: cwl_runner/cwl_startup.sh [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+  for url in "${URL[@]}"; do
```
Comment (mbookman) [inline cwl_runner/cwl_startup.sh]: Change `url` to `URL`

File: cwl_runner/cwl_startup.sh [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+    url=$(echo ${url} | tr -d '"')  # Remove quotes
+    url_local="${INPUT_FOLDER}/$(dirname ${url//:\//})"
```
Comment (mbookman) [inline cwl_runner/cwl_startup.sh]: Change `url_local` to `URL_LOCAL`

File: cwl_runner/cwl_runner.sh (hunk 17 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: The --metadata can't contain spaces -- indenting breaks the gcloud parsing.

File: cwl_runner/cwl_startup.sh (hunk 26 above)
Comment (jbingham) [inline cwl_runner/cwl_startup.sh]: Done

File: cwl_runner/cwl_startup.sh (hunk 27 above)
Comment (jbingham) [inline cwl_runner/cwl_startup.sh]: Done

File: cwl_runner/cwl_startup.sh (hunk 25 above)
Comment (jbingham) [inline cwl_runner/cwl_startup.sh]: Done

File: cwl_runner/cwl_startup.sh (hunk 24 above)
Comment (jbingham) [inline cwl_runner/cwl_startup.sh]: Done

File: cwl_runner/cwl_shutdown.sh (hunk 23 above)
Comment (jbingham) [inline cwl_runner/cwl_shutdown.sh]: Done

File: cwl_runner/cwl_shutdown.sh (hunk 22 above)
Comment (jbingham) [inline cwl_runner/cwl_shutdown.sh]: Done

File: cwl_runner/cwl_shutdown.sh (hunk 21 above)
Comment (jbingham) [inline cwl_runner/cwl_shutdown.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 20 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 19 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 18 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 16 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 15 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 14 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 12 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 11 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/cwl_runner.sh (hunk 10 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Done

File: cwl_runner/README.md (hunk 9 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 8 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 7 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 5 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 6 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 4 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 3 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 2 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/README.md (hunk 1 above)
Comment (jbingham) [inline cwl_runner/README.md]: Done

Comment (jbingham) [inline cwl_runner/README.md]: Done

File: cwl_runner/cwl_runner.sh (hunk 13 above)
Comment (jbingham) [inline cwl_runner/cwl_runner.sh]: Added nounset and moved both a bit earlier. Something is add about the HELP_MESSAGE assignment though, so have to set errexit *after* that point.

**Comment by jbingham** [comment]
All updates in. See the comment about errexit/nounset. 

Rerunning GDC workflow to confirm no typos introduced.

**Comment by jbingham** [comment]
Test run completed successfully. Ready for next review.

**Comment by jbingham** [comment]
Tiny tweak: one declare replaced with readonly, and one var uppercased. OK?

**Pull request merged**